include/picongpu/simulation/stage/AtomicPhysics.hpp

```diff
@@ -0,0 +1,604 @@
+                        true>>{}(mappingDesc, "LocalRejectionProbabilityCacheField");
+            }
+
+            //! print local time remaining to console, debug only
+            HINLINE static void printTimeRemaingToConsole(picongpu::MappingDesc const mappingDesc)
+            {
+                picongpu::particles::atomicPhysics::stage::DumpSuperCellDataToConsole<
+                    picongpu::particles::atomicPhysics::localHelperFields::LocalTimeRemainingField<
+                        picongpu::MappingDesc>,
+                    picongpu::particles::atomicPhysics::localHelperFields::PrintTimeRemaingToConsole>{}(
+                    mappingDesc,
+                    "LocalTimeRemainingField");
+            }
+
+            //! print local time step to console, debug only
+            HINLINE static void printTimeStepToConsole(picongpu::MappingDesc const mappingDesc)
+            {
+                picongpu::particles::atomicPhysics::stage::DumpSuperCellDataToConsole<
+                    picongpu::particles::atomicPhysics::localHelperFields::LocalTimeStepField<picongpu::MappingDesc>,
+                    picongpu::particles::atomicPhysics::localHelperFields::PrintTimeStepToConsole>{}(
+                    mappingDesc,
+                    "LocalTimeStepField");
+            }
+
+        public:
+            AtomicPhysics() = default;
+
+            //! atomic physics stage sub-stage calls
+            void operator()(picongpu::MappingDesc const mappingDesc, uint32_t const currentStep) const
+            {
+                // debug only
+                using ForEachIonSpeciesDumpRateCacheToConsole = pmacc::meta::ForEach<
+                    AtomicPhysicsIonSpecies,
+                    particles::atomicPhysics::stage::DumpRateCacheToConsole<boost::mpl::_1>>;
+
+                pmacc::DataConnector& dc = pmacc::Environment<>::get().DataConnector();
+
+                // TimeRemainingSuperCellField
+                auto& localTimeRemainingField = *dc.get<S_TimeRemainingField>("LocalTimeRemainingField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutTimeRemaining
+                    = localTimeRemainingField.getGridLayout().sizeWithoutGuardND();
+
+                // FoundUnboundIonSuperCellField
+                auto& localFoundUnboundIonField = *dc.get<S_FoundUnboundField>("LocalFoundUnboundIonField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutFoundUnbound
+                    = localFoundUnboundIonField.getGridLayout().sizeWithoutGuardND();
+
+                // ElectronHistogramOverSubscribedSuperCellField
+                auto& localElectronHistogramOverSubscribedField
+                    = *dc.get<S_OverSubscribedField>("LocalElectronHistogramOverSubscribedField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutOverSubscription
+                    = localElectronHistogramOverSubscribedField.getGridLayout().sizeWithoutGuardND();
```

Comment:
REFACTOR: For these variables it would have been nice to have a
```C++
struct LocalField
    field;
    layout;
    
    LocalField(&dc) : field{*dc.get...}, layout{field.getGridLayout()...}
};
```
to have these 14 lines down to three.

REQUEST: Also: Remove comments. They are redundant with the (quite good!) naming.

Reply:
comments removed